certora/specs/ERC20Invariants.spec

```diff
@@ -0,0 +1,138 @@
+    require ghostDelegatee[account] == delegatee;
+}
+
+// Slot is DelegationTokenStorage._delegatee.
+hook Sstore (slot 0x669be2f4ee1b0b5f3858e4135f31064efe8fa923b09bf21bf538f64f2c3e1100)[KEY address account] address delegatee (address delegateeOld) {
+    mathint changeOfAccountVotesForA;
+    // Track delegation changes from the parameterized address.
+    if (delegateeOld == A && delegatee != A) {
+        require changeOfAccountVotesForA == - ghostBalances[account];
+    // Track delegation changes to the pramaeterized address.
+    } else if (delegateeOld != A && delegatee == A) {
+        require changeOfAccountVotesForA == ghostBalances[account];
+    } else {
+        require changeOfAccountVotesForA == 0;
+    }
+    // Update partial sums for x > to_mathint(account)
+    havoc sumsOfVotes assuming
+        forall mathint x. sumsOfVotes@new[x] ==
+        sumsOfVotes@old[x] + (to_mathint(account) < x ? changeOfAccountVotesForA : 0);
+    // Update ghost copy of DelegationTokenStorage._delegatee.
+    ghostDelegatee[account] = delegatee;
+}
+
+// Ghost variable to hold the sum of balances.
+ghost mapping(mathint => mathint) sumOfBalances {
+    init_state axiom forall mathint addr. sumOfBalances[addr] == 0;
+}
+
+// Ghost copy of ERC20Storage._balances for quantification.
+ghost mapping(address => uint256) ghostBalances {
+    init_state axiom forall address account. ghostBalances[account] == 0;
+}
+
+// Slot is ERC20Storage._balances slot.
+hook Sload uint256 balance (slot 0x52c63247e1f47db19d5ce0460030c497f067ca4cebf71ba98eeadabe20bace00)[KEY address account] {
+    require ghostBalances[account] == balance;
+}
+
+// Slot is ERC20Storage._balances slot
+hook Sstore (slot 0x52c63247e1f47db19d5ce0460030c497f067ca4cebf71ba98eeadabe20bace00)[KEY address account] uint256 newValue (uint256 oldValue) {
+    // Update partial sum of balances, for x > to_mathint(account)
+    // Track balance changes in balances.
+    havoc sumOfBalances assuming
+        forall mathint x. sumOfBalances@new[x] ==
+            sumOfBalances@old[x] + (to_mathint(account) < x ? newValue - oldValue : 0);
+    // Update partial sums of votes delegated to the parameterized address, for x > to_mathint(account)
+    // Track balance changes when the delegatee is the parameterized address.
+    if (ghostDelegatee[account] == A) {
+        havoc sumsOfVotes assuming
+            forall mathint x. sumsOfVotes@new[x] ==
+                sumsOfVotes@old[x] + (to_mathint(account) < x ? newValue - oldValue : 0);
+    }
+    // Update ghost copy of ERC20Storage._balances.
+    ghostBalances[account] = newValue;
+}
+
+invariant sumOfBalancesStartsAtZero()
+    sumOfBalances[0] == 0;
+
+invariant sumOfBalancesGrowsCorrectly()
+    forall address addr. sumOfBalances[to_mathint(addr) + 1] ==
+        sumOfBalances[to_mathint(addr)] + ghostBalances[addr];
+
+invariant sumOfBalancesMonotone()
```

Comment:
trivial with `sumOfBalancesGrowsCorrectly` as ghostBalances are positive right? do we need to keep it anyway?

Reply:
I assume it's trivial for us, but not easy to find by the prover. Notably, proving `sumOfBalancesMonotone` from `sumOfBalancesGrowsCorrectly` requires to do an induction, which provers usually struggle with. Not sure if it's necessary for the following proofs though